shepherd/shepherd/shepherd.py

```diff
@@ -188,65 +206,105 @@ def _slaughter_sheep(self, sheep_id: str) -> None:
 
         :param sheep_id: sheep id to be fed
         """
+        sheep = self._get_sheep(sheep_id)
+
         while True:
-            sheep = self._get_sheep(sheep_id)
             job_id = await sheep.jobs_queue.get()
-            logging.info('Preparing working directory for job `%s` on `%s`', job_id, sheep_id)
-
-            # prepare working directory
-            working_directory = create_clean_dir(path.join(sheep.sheep_data_root, job_id))
-            await self._storage.pull_job_data(job_id, working_directory)
-            create_clean_dir(path.join(working_directory, OUTPUT_DIR))
-
-            # update the job status
-            status = self._job_status[job_id]
-            status.status = JobStatus.PROCESSING
-            status.processing_started_at = datetime.utcnow()
-            await self._job_status_update_queue.enqueue_task(self._storage.set_job_status(job_id, status.copy()))
 
-            # (re)start the sheep if needed
-            model = status.model
-            if model.name != sheep.model_name or model.version != sheep.model_version or not sheep.running:
-                logging.info('Job `%s` requires model `%s:%s` on `%s`', job_id, model.name, model.version, sheep_id)
-                # we need to wait for the in-progress jobs which are already in the socket
-                async with self.job_done_condition:
-                    await self.job_done_condition.wait_for(lambda: len(sheep.in_progress) == 0)
-                self._slaughter_sheep(sheep_id)
+            try:
+                # wait until working directory is prepared
                 try:
-                    self._start_sheep(sheep_id, model.name, model.version)
-                except SheepConfigurationError as sce:
-                    error = ErrorModel({
-                        'message': 'Failed to start sheep for this job ({})'.format(str(sce))
-                    })
-
-                    logging.error('Sheep `%s` encountered error when processing job `%s`: %s', sheep_id, job_id, error.message)
-                    await self._report_job_failed(job_id, error, sheep)
-                    continue
-                except Exception as ex:
-                    error = ErrorModel({
-                        'message': '`{}` thrown when starting sheep `{}` for job `{}`'.format(str(ex), sheep_id, job_id),
-                        'exception_type': str(type(ex)),
-                        'exception_traceback': str(traceback.format_tb(ex.__traceback__))
-                    })
-
-                    await self._report_job_failed(job_id, error, sheep)
-                    logging.exception("Error encountered when starting sheep `%s` for job `%s`", sheep_id, job_id)
+                    await self._prepare_job_dir_futures[job_id]
+                    del self._prepare_job_dir_futures[job_id]
+                except KeyError:
+                    logging.critical(f"Task that should have pulled data for job `{job_id}` mysteriously disappeared")
```

Comment:
This may happen anytime a sheep is purged with `health_check`, I do not think that it deserves the `critical` level...